File: parser/src/main/java/com/labuda/gdlunch/parser/zomato/ZomatoParser.java
package com.labuda.gdlunch.parser.zomato;

import com.labuda.gdlunch.entity.DailyMenu;
import com.labuda.gdlunch.entity.MenuItem;
import com.labuda.gdlunch.entity.Restaurant;
import com.labuda.gdlunch.parser.AbstractRestaurantWebParser;
import com.labuda.gdlunch.parser.DailyParser;
import java.time.LocalDate;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Basic Zomato parser
 */
public class ZomatoParser extends AbstractRestaurantWebParser implements DailyParser {

    /**
     * Logger
     */
    private final static Logger log = LoggerFactory.getLogger(ZomatoParser.class);

    /**
     * Zomato REST client
     */
    private ZomatoRestClient zomatoRestClient = new ZomatoRestClient();

    /**
     * Constructor
     *
     * @param restaurant restaurant
     */
    public ZomatoParser(Restaurant restaurant) {
        super(restaurant);
    }

    /**
     * Obtains daily menu
     *
     * @return daily menu as JSON
     */
    protected JSONObject getDailyMenu() {
        return zomatoRestClient.getDailyMenu(restaurant.getParserUrl());
    }

    @Override
    public DailyMenu parse() {
        DailyMenu result = new DailyMenu();
        result.setDate(LocalDate.now());
        result.setRestaurant(restaurant);

        try {
            JSONObject dailyMenu = getDailyMenu();
            JSONArray dishes = dailyMenu.getJSONArray("daily_menus").getJSONObject(0)
                    .getJSONObject("daily_menu").getJSONArray("dishes");

            for (int i = 0; i < dishes.length(); i++) {
                JSONObject dish = dishes.getJSONObject(i).getJSONObject("dish");

                MenuItem menuItem = new MenuItem();
                menuItem.setName(dish.getString("name").trim());
                menuItem.setPrice(parsePrice(dish.getString("price")));

                result.getMenu().add(menuItem);
            }
        } catch (JSONException e) {
            log.error("Parsing of: " + restaurant + "has failed.", e);
        }

        addTranslations(result);
        return result;
    }
}

File: src/main/kotlin/gdlunch/parser/zomato/ZomatoParser.kt
package gdlunch.parser.zomato

import com.labuda.gdlunch.parser.AbstractRestaurantWebParser
import com.labuda.gdlunch.parser.DailyParser
import com.labuda.gdlunch.repository.entity.DailyMenu
import com.labuda.gdlunch.repository.entity.MenuItem
import com.labuda.gdlunch.repository.entity.Restaurant
import java.time.LocalDate

/**
 * Basic Zomato parser for daily menus
 */
class ZomatoParser(restaurant: Restaurant) : AbstractRestaurantWebParser(restaurant), DailyParser {

    /**
     * REST client for communication with Zomato
     */
    private val zomatoRestClient = ZomatoRestClient()

    override fun parse(): DailyMenu = DailyMenu(
            LocalDate.now(),
            restaurant,
            getListOfDishes()
    )

    /**
     * Transforms the Zomato response to a list menu items
     */
    private fun getListOfDishes(): List<MenuItem> {
        val dailyMenus = zomatoRestClient.getDailyMenu(restaurant.parserUrl).dailyMenus
        return if (dailyMenus.isNotEmpty()) {
            dailyMenus[0].dailyMenu.dishes.map { MenuItem(it.dish.name, parsePrice(it.dish.price)) }
        } else {
            listOf(MenuItem())
        }
    }
}
